pkg/proxy/iptables/proxier.go

```diff
@@ -1351,7 +1342,7 @@ func (proxier *Proxier) syncProxyRules() {
 		endpoints = endpoints[:0]
 		endpointChains = endpointChains[:0]
 		var endpointChain utiliptables.Chain
-		for _, ep := range readyEndpoints {
+		for _, ep := range allEndpoints {
```

Comment:
> This is going to make Chains for all endpoints, even if we choose not to use them. Is that what we want?

Yes, all chains need to exist so we can use them when needed, but the iptables probablistic loadbalancing will not include chains we don't want.

Reply:
Yes, this does increase total lines in iptables-restore, but it doesn't actually change the number of rules evaluated, since the service chain decides which endpoint chain to jump to based on probability, and that never includes terminating endpoints unless they are the only endpoints remaining. 

So IMO as long as we are not adding more rules to be evaluated and assuming there aren't a huge number of unready endpoints compared to ready endpoints, this seems safe.

Reply:
^ although being honest I have not run extensive tests to prove this :)

Reply:
I've found the code, if nobody beats me I'll run it with and without this PR https://github.com/kubernetes/kubernetes/commit/d64d3da2ed1d34f09b2acee38b0e5c93118048e0

Reply:
hmm, I should be doing something wrong, I have the same number of iptables rules in this PR, in master and in 1.21.1

https://gist.github.com/aojea/1308f16ad611d76d64b1eafd7bfc9c88#file-experiment-md

Reply:
oh, so this only increases the number of lines because we also create rules for the "unready endpoints", I don't know how to address that honestly, what is the proportion of unready endpoints and the average time they last?

Reply:
> what is the proportion of unready endpoints and the average time they last?

I would think that the # of unready endpoints is typically much lower than ready endpoints. But in certain failure scenarios, I can see this number being higher. It will always vary depending on the cluster and the workloads there.

Reply:
> This is what I am concerned about and I simply don't have data that allows us to make the decision.

absolutely agree, this is going to have a totally different behavior depending on the cluster and the workloads, and I can't see how we can know if we are going to regress on performance on 80%, 50%, 15%, ... of the clusters ...

> The technical solution would be to overhaul some of this code to refcount endpoints and elide any that we don't need. That's not a trivial amount of work (I think) but I don't know how to be "safe" without it.

.. we can always make this opt-in with a well-known annotation

Reply:
> .. we can always make this opt-in with a well-known annotation

Personally against using annotations for things like this, especially if we are trying to do the right things automatically. I think in most cases, falling back to terminating is the right behavior (I could be totally wrong here though).

> And I am not saying "go do it" but "let's convince ourselves that we do or don't need to"..

Quite frankly I'm not sure how to measure this accurately but I'm definitely open to hearing suggestions and trying to run different tests that give us some meaningful data. In the meantime, I can also look into how much extra  work it would be to pre-calculate whenever an endpoint chain should exist or not. I agree it's probably not a trivial amount of work.

Reply:
> Personally against using annotations for things like this, especially if we are trying to do the right things automatically. I think in most cases, falling back to terminating is the right behavior (I could be totally wrong here though).
> 
> >

agree, just was trying to provide alternatives to move forward :smile: